src/main/java/org/jenkins/tools/test/model/plugin_metadata/LocalCheckoutMetadataExtractor.java

```diff
@@ -0,0 +1,106 @@
+package org.jenkins.tools.test.model.plugin_metadata;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.jenkins.tools.test.exception.MetadataExtractionException;
+import org.jenkins.tools.test.exception.PluginSourcesUnavailableException;
+import org.jenkins.tools.test.exception.PomExecutionException;
+import org.jenkins.tools.test.maven.ExternalMavenRunner;
+import org.jenkins.tools.test.maven.MavenRunner;
+import org.jenkins.tools.test.model.PluginCompatTesterConfig;
+import org.jenkins.tools.test.model.plugin_metadata.PluginMetadata.Builder;
+
+public class LocalCheckoutMetadataExtractor {
+
+    private static final Logger LOGGER = Logger.getLogger(LocalCheckoutMetadataExtractor.class.getName());
+
+    // artifactId\tversion\t\directory
+    private static Pattern p = Pattern.compile("(?<id>[^\t]+)\t(?<version>[^\t]+)\t(?<path>[^\t]+)");
+
+    public static List<PluginMetadata> extractMetadata(File localCheckoutDir, PluginCompatTesterConfig config)
+            throws MetadataExtractionException {
+        File log = new File(config.getWorkingDir(), "local-metadata.log");
+        MavenRunner runner =
+                new ExternalMavenRunner(config.getExternalMaven(), config.getMavenSettings(), config.getMavenArgs());
+        Set<String> excludedPlugins = config.getExcludePlugins();
+        Set<String> includedPlugins = config.getIncludePlugins();
+        try {
+            runner.run(
+                    Map.of("output", log.getAbsolutePath()),
+                    localCheckoutDir,
+                    null,
+                    null,
+                    "-q",
+                    // TODO to switch to release version
+                    // https://github.com/jenkinsci/maven-hpi-plugin/pull/463
+                    "org.jenkins-ci.tools:maven-hpi-plugin:3.42-rc1408.71cefb_fc63b_d:list-plugins",
```

Comment:
I think we should override this version only if necessary (i.e., only if the plugin is using a version older than the minimum required version). There is code elsewhere in this repository that implements this already.